src/utils/saml.ts

```diff
@@ -1059,24 +1072,30 @@ export const getPreValidateResponse = (
           )
       )
       .chain(_ =>
-        fromOption(new Error("Assertion element must be present"))(
-          fromNullable(
-            _.Response.getElementsByTagNameNS(
+        fromPredicate<Error, { IssueInstant: Date; Response: Element }>(
+          predicate =>
+            predicate.Response.getElementsByTagNameNS(
               SAML_NAMESPACE.ASSERTION,
               "Assertion"
-            ).item(0)
-          )
-        ).map(Assertion => ({
-          Assertion,
-          ..._
-        }))
+            ).length === 1,
+          _1 => new Error("SAML Response must have only one Assertion element")
+        )(_).chain(_1 =>
+          fromOption(new Error("Assertion element must be present"))(
+            fromNullable(
+              _1.Response.getElementsByTagNameNS(
+                SAML_NAMESPACE.ASSERTION,
+                "Assertion"
+              ).item(0)
+            )
```

Comment:
isn't this implied by the previous check?

Reply:
This code is only for type check. We need a cast like `.item(0) as Element` to avoid the `fromNullable`

Reply:
can you add this as a comment?

Reply:
I've changed the predicate to discriminate 2 different errors:
- multiple Responses or Assertions
- missing Response or Assertion

Fixed in a729df3